test/Spago/Build.purs

```diff
@@ -99,65 +101,212 @@ spec = Spec.around withTempDir do
+                , "import Effect.Console as Console"
+                , "main = unit"
+                ]
+              FS.writeTextFile (Path.concat [ "test", "Test", "Main.purs" ]) $ Array.intercalate "\n"
+                [ "module Test.Main where"
+                , "import Prelude"
+                , "import Data.Maybe as Maybe"
+                , "import Effect as Effect"
+                , "import Effect.Console as Console"
+                , "import Control.Alt as Alt"
+                , "main = unit"
+                ]
+              -- get rid of "Compiling ..." messages and other compiler warnings
+              spago [ "build" ] >>= shouldBeSuccess
+
+          Spec.it "passed --pedantic-packages CLI flag" \{ spago, fixture } -> do
+            setupTestTransitiveTests spago
+            spago [ "build", "--pedantic-packages" ] >>= shouldBeFailureErr (fixture "check-direct-import-transitive-dependency-test.txt")
+
+          Spec.it "package config has 'pedantic_packages: true'" \{ spago, fixture } -> do
+            setupTestTransitiveTests spago
+            addPedanticPackagesToTest
+            spago [ "build" ] >>= shouldBeFailureErr (fixture "check-direct-import-transitive-dependency-test.txt")
+
+      Spec.describe "fails with warnings about unused dependencies" do
+
+        Spec.describe "in a source package when" do
+
+          let
+            setupSrcUnusedDeps spago installConsoleAndEffect = do
+              spago [ "init", "--name", "7368613235362d2f444a2b4f56375435646a59726b53586548" ] >>= shouldBeSuccess
+              when installConsoleAndEffect do
+                spago [ "install", "--test-deps", "console", "effect" ] >>= shouldBeSuccess
+              FS.writeTextFile (Path.concat [ "src", "Main.purs" ]) "module Main where\nimport Prelude\nmain = unit"
+              -- get rid of "Compiling ..." messages and other compiler warnings
+              spago [ "build" ] >>= shouldBeSuccess
 
-        Spec.it "passing --pedantic-packages CLI flag" \{ spago, fixture } -> do
-          setupSrcUnusedDeps spago
-          spago [ "build", "--pedantic-packages" ] >>= shouldBeFailureErr (fixture "check-unused-dependency.txt")
+          Spec.it "passing --pedantic-packages CLI flag" \{ spago, fixture } -> do
+            setupSrcUnusedDeps spago true
+            spago [ "build", "--pedantic-packages" ] >>= shouldBeFailureErr (fixture "check-unused-dependency.txt")
 
-        Spec.it "package config has 'pedantic_packages: true'" \{ spago, fixture } -> do
-          setupSrcUnusedDeps spago
-          addPedanticPackagesToSrc
-          spago [ "build" ] >>= shouldBeFailureErr (fixture "check-unused-dependency.txt")
+          Spec.it "package config has 'pedantic_packages: true'" \{ spago, fixture } -> do
+            setupSrcUnusedDeps spago false
+            addPedanticPackagesToSrc
+            spago [ "build" ] >>= shouldBeFailureErr (fixture "check-unused-dependency.txt")
```

Comment:
I think we have another issue here:
- the source dependencies have `console` and `effect`
- ...but the sources don't use them
- the test dependencies don't have them
- ...but the test sources _do_ use them

We correctly complain about them being unused (because we have the flag set to true in the package), but since the flag is not set in tests we don't say anything there. 
However removing them would make the build fail, since they are used in tests. Not sure how to disentangle this one.

Reply:
The options forward are:
1. Stop making `test` depend on all the packages declared by the source package implicitly. Rather, make it explicit. A `test` package only implicitly depends on its source package; that's it. While this simplifies this particular situation, it complicates `spago install` and `spago uninstall` situations. But hey, isn't that what the `--pedantic-packages=autofix` idea is for?
2. Add a `purs graph` call that allows one to see which dependencies in the source package are actually used and have the test package only implicitly depend on those used dependencies. While this resolves this issue, it'll be confusing because a `spago uninstall` can make a future `spago build` not work as (correct me if I'm wrong) `spago build -p foo` build both the source and test packages of `foo`.
3. Add another check for this situation. If source declares unused deps that tests does not declare but which uses, then include it as an error. Since we're already going to be reporting that the source package has unused dependencies and thereby failing the build, we may also state the above issue.

Reply:
Any new thoughts on this? Or are you still thinking about it?

Reply:
I'm still thinking about the overall shape of this PR, but I think I also prefer option 3 here

Reply:
Yeah, I think option 1 is a possibility but definitely uncomfortable, so I'd like to reach for it only in case we can't make anything else work.

As for Option 2, I am not sure how it resolves the issue - I think `spago uninstall` is supposed to be able to break the build, so that part is fine. The test package already implicitly depends on the packages, the issue is that we're not detecting if we are slating these implicit dependencies for removal.

Option 3 makes sense, and I'd go even further in the source+test packages integration, and add another kind of "error" that we return - on top of "unused" and "undeclared transitive" - which is "unused in source but used in test" (I'm sure we can find a more compact naming 😄). So in the same graph check we'd take the unused packages for the source package, and see if they'd be used in the test package. If yes, we add `spago install --test-deps` instructions for them, which should counterbalance the `spago uninstall` ones.